common/migration/readme.md

```diff
@@ -0,0 +1,132 @@
+
+This library helps by allowing the code to perform this switch at runtime
+and handles some of the annoying parts of this switching, such as ensuring that 
+the new codepath is not going to wreck things if it's either too slow, panicing or 
+otherwise misbehaving. It's not perfect, there's still ways to break the system with 
+poorly behaved code, but ignoring the context timeout panics at least should be handled.
+
+How does it work? The intuition is that for each migration case a new
+instance would be created of the migration lib... 
+
+```
+// assuming a fictitous user controller:
+
+func NewUserController(p params) {
+
+    // a custom comparison for example's sake
+    userComparisonFn := func(
+        log log.Logger,
+        scope metrics.Scope,
+        activeFlow *User,
+        activeErr error,
+        backgroundFlow *User,
+        backgroundErr error) (isEqual bool) {
+	
+	    if activeFlow.UserID == backgroundFlow.UserID {
+	        return true 
+	    }
+	    
+	    // the data isn't the same, write some complaining logs
+	    log.Error("The data being returned from the flows isn't equal")
+	    
+	    return false
+	}
+
+    migration := NewDualReaderWithCustomComparisonFn[*User](
+		dynamicconfig.UserMigrationRollout, // create a string based dynamic config for this rollout
+		func(_ ...dynamicconfig.Filter) bool { return true } // always compare results
+		p.log,
+		p.scope,
+		time.Second * 15, // hard cutoff for any background shadow requests in case the go longer than the ctx
+		userComparisonFn)
+
+    return Controller{
+        migration: migration
+        ... 
+    }
+}
+
+```
+
+... and for the new and old codepaths, they should be able to fit in a function
+
+```
+    
+    func (c *controller) GetUserController (ctx context.Context, userID string) (*User, error) {
+
+        oldFlow := func(ctx context.Context) (*getUserResponse, error){
+            // .. do whatever is necessary to setup the params
+            c.getLegacyUserFlow(ctx, requestParams)
+        }
+        
+        newFlow := func(ctx context.Context) (*getUserResponse, error){
+            // create request...
+            c.getWebscaleUsers(ctx, requestParams, filters)
+        }
+        
+        constraints := migration.Contraints{
+            Operation: "get-user-flow",
+            Domain: "some-test-domain",
+        }
+
+        user, err := c.migration.ReadAndReturnActive(ctx, constraints, oldFlow, newFlow)
+        
+        // so the user/err here is whatever's in the 'active' flow. That's controlled
+        // by the dynamic config mentioned earlier.
+        return user, err
+}
+```
+
+#### Rollout sequence 
+
+With the rollout dynamic config, the following values being returned will controll the behaviour
+of the library, with a hopefully fairly self-evident behaviour. By default it will only call the old flow.
+
+Expected rollout states:
+- `only-call-old`
+- `call-both-return-old`
+- `call-both-return-new`
+- `only-call-new`
```

Comment:
since we have or will have multiple of these "configure the shadow rollout" things, tbh I wonder if it's worth coming up with a standard syntax for them.

e.g. the global ratelimiter uses `local` / `global` / `active-shadow-other`.

not that I think this is blocking or anything.  just a nice-to-have / if you have ideas.

Reply:
yeah, this part I'm super not thrilled about and totally open to suggestions. Using strings is error prone, but at least easy to understand. I used enums before, but they're hard to undrestand when in another system like flipr.

Reply:
yea :\  Go enums offer absurdly weak guarantees anyway, so they don't really eliminate any error paths.

structured strings as input makes sense I think, since it's coming from the outside - a dynamic config store of some kind.  it's not truly hardcoded, we still have to parse something.  but I do like having a type to mark "this has been validated" rather than throwing around strings.

one option I considered but only half built before simplifying: could parse it into a struct that has "call X" and "use X" methods.  then you can bake fallback behavior into the responses, and not have to deal with a `default:` branch in switch statements or something.